src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java

```diff
@@ -19,63 +27,186 @@
+                "", "", 0);
+        private ConcurrentMap<String, ConcurrentMap<String, MinimalPullRequest>> pullRequests;
+
+        public RepositoryStore(
+                ConcurrentMap<String, ConcurrentMap<String, MinimalPullRequest>> pullRequests) {
+            this.pullRequests = pullRequests;
+        }
+
+        public void closePullRequest(String fromBranch, String toBranch) {
+            pullRequests.getOrDefault(fromBranch, new ConcurrentHashMap<>()).getOrDefault(toBranch, CLOSED_PR)
+                    .setState(BitbucketPullState.DELETED);
+        }
+
+        public AtomicLong findOldestUpdate() {
+            AtomicLong oldestUpdate = new AtomicLong(0);
+            pullRequests.values().stream().forEach(fromBranch -> {
+                fromBranch.values().forEach(pr -> {
+                    if (oldestUpdate.get() == 0) {
+                        oldestUpdate.set(pr.getUpdatedDate());
+                    }
+                    if (pr.getUpdatedDate() < oldestUpdate.get()) {
+                        oldestUpdate.set(pr.getUpdatedDate());
+                    }
+                });
+            });
+            return oldestUpdate;
+        }
+
+        public MinimalPullRequest getPullRequest(String fromBranch, String toBranch) {
+            return pullRequests.getOrDefault(fromBranch, new ConcurrentHashMap<>()).get(toBranch);
+        }
+
+        public boolean hasOpenPullRequests(String branchName) {
+            return pullRequests.getOrDefault(branchName, new ConcurrentHashMap<>()).values().stream().filter(
+                    pr -> pr.getState() == BitbucketPullState.OPEN).findFirst().isPresent();
+        }
+
+        public boolean hasPR() {
+            AtomicBoolean result = new AtomicBoolean(false);
+            pullRequests.values().forEach(toMap -> {
+                if (!toMap.isEmpty()) {
+                    result.set(true);
+                }
+            });
+            return result.get();
+        }
+
+        public void removeClosedPullRequests(long date) {
+            pullRequests.values().forEach(toMap -> {
+                toMap.keySet().forEach(ref -> {
+                    toMap.compute(ref, (toBranch, pr) -> {
+                        if (pr.getUpdatedDate() < date) {
+                            if (pr.getState() != BitbucketPullState.OPEN) {
+                                return null;
+                            }
+                        }
+                        return pr;
+                    });
+                });
+            });
+        }
+
+        public void updatePullRequest(MinimalPullRequest pr) {
+            pullRequests.compute(pr.getFromRefDisplayId(), getUpdatePrBiFunction(pr));
+        }
+
+        private BiFunction<String, ConcurrentMap<String, MinimalPullRequest>,
+                ConcurrentMap<String, MinimalPullRequest>> getUpdatePrBiFunction(MinimalPullRequest pr) {
+            return (key, map) -> {
```

Comment:
maybe some better names? `key` and `map` feels very generic :slightly_smiling_face: